NDNc/pipeline/pipeline-interests-aimd.cpp

```diff
@@ -0,0 +1,187 @@
+                                   ndn::lp::PitToken &&pitToken) {
+    auto pitEntry = getPITTokenValue(std::move(pitToken));
+
+    if (m_pit->find(pitEntry) == m_pit->end()) {
+        std::cout << "DEBUG: unexpected Data packet dropped\n";
+        return;
+    }
+
+    if (data->getCongestionMark()) {
+        decreaseWindow();
+        std::cout << "Debug: ECN received\n";
+    }
+
+    enqueueData(std::move(data)); // Enqueue Data into Response queue
+    m_pit->erase(pitEntry);
+    increaseWindow();
+}
+
+void PipelineInterestsAimd::onNack(std::shared_ptr<ndn::lp::Nack> &&nack,
+                                   ndn::lp::PitToken &&pitToken) {
+
+    auto pitEntry = getPITTokenValue(std::move(pitToken));
+    std::cout << "INFO: received NACK with reason " << nack->getReason()
+              << "\n";
+
+    switch (nack->getReason()) {
+    case ndn::lp::NackReason::DUPLICATE: {
+        auto interest = getWireDecode(m_pit->at(pitEntry).interest);
+        interest->refreshNonce();
+
+        uint64_t newPITEntry = m_rdn->get();
+        auto lifetime = interest->getInterestLifetime().count();
+
+        auto pendingInterest = PendingInterest(
+            std::move(getWireEncode(std::move(interest), newPITEntry)),
+            newPITEntry, lifetime);
+
+        m_requestQueue.enqueue(std::move(pendingInterest));
+        m_pit->erase(pitEntry);
+        break;
+    }
+    default:
+        enqueueData(nullptr); // Enqueue null to mark error
+        m_pit->erase(pitEntry);
+        break;
+    }
+}
+
+void PipelineInterestsAimd::onTimeout() {
+    while (!m_queue->empty()) {
+        if (m_pit->count(m_queue->front()) == 0) {
+            // Pop entries that were already satisfied with success
+            m_queue->pop();
+            continue;
+        }
+
+        auto entry = m_pit->find(m_queue->front());
+        if (!entry->second.isExpired()) {
+            return;
+        }
+
+        m_queue->pop();
+
+        auto interest = getWireDecode(entry->second.interest);
+        interest->refreshNonce();
+
+        decreaseWindow();
+
+        auto newPITEntry = m_rdn->get();
+        auto lifetime = interest->getInterestLifetime().count();
+
+        auto pendingInterest = PendingInterest(
+            std::move(getWireEncode(std::move(interest), newPITEntry)),
+            newPITEntry, lifetime);
+        pendingInterest.nTimeout = entry->second.nTimeout + 1;
+
+        if (pendingInterest.nTimeout < 8) {
+            m_pit->erase(entry);
+            m_requestQueue.enqueue(std::move(pendingInterest));
+        } else {
+            enqueueData(nullptr); // Enqueue null to mark error
+            m_pit->erase(entry);
+        }
+    }
+}
+void PipelineInterestsAimd::decreaseWindow() {
+    ndn::time::steady_clock::time_point now = ndn::time::steady_clock::now();
+    if (now - m_lastDecrease < MAX_RTT) {
+        return;
+    }
+    std::cerr << "window decrease at " << m_windowSize << '\n';
```

Comment:
please specify the type of message you want this to be: DEBUG/INFO/TRACE. I am in the process of introducing a logger this week and it would really help me when updating everything